src/data_structures/bwt.rs

```diff
@@ -123,11 +123,46 @@ impl Occ {
     /// Get occurrence count of symbol a in BWT[..r+1].
     /// Complexity: O(k).
     pub fn get(&self, bwt: &BWTSlice, r: usize, a: u8) -> usize {
-        let i = r / self.k;
-
-        let count = bwt[(i * self.k) + 1..r + 1].iter().filter(|&&c| c == a).count();
+        // NOTE:
+        //
+        // Retrieving byte match counts in this function is critical to the performance of FM Index.
+        //
+        // The below manual count code is roughly equivalent to:
+        // ```
+        // let count = bwt[(i * self.k) + 1..r + 1].iter().filter(|&&c| c == a).count();
+        // self.occ[i][a as usize] + count
+        // ```
+        //
+        // But there are a couple of reasons to do this manually:
+        // 1) As of 2016, versions of rustc/LLVM vectorize this manual loop more reliably
+        //    than the iterator adapter version.
+        // 2) Manually accumulating the byte match count in a single chunk can allows
+        //    us to use a `u32` for that count, which has faster arithmetic on common arches.
+        //    This does necessitate storing `k` as a u32.
+        //
+        // See the conversation in these issues for some of the history here:
+        //
+        // https://github.com/rust-bio/rust-bio/pull/74
+        // https://github.com/rust-bio/rust-bio/pull/76
+
+        // self.k is our sampling rate, so find our last sampled checkpoint
+        let i = r / self.k as usize;
+        let checkpoint = self.occ[i][a as usize];
+
+        // find the portion of the BWT past the checkpoint which we need to count
+        let start = (i * self.k as usize) + 1;
+        let end = r + 1;
+
+        // count all the matching bytes b/t the closest checkpoint and our desired lookup
+        let mut count: u32 = 0;
+        for &x in &bwt[start..end] {
+            if x == a {
+                count += 1;
```

Comment:
From running benches on my 2016 MBP.

The current PR:

```
test search_index_seeds ... bench:      33,975 ns/iter (+/- 1,272)
test search_index_seeds ... bench:      33,912 ns/iter (+/- 3,118)
test search_index_seeds ... bench:      34,024 ns/iter (+/- 963)

average: ~33970 ns/iter

with RUSTFLAGS="-C target-cpu=native"

test search_index_seeds ... bench:      33,715 ns/iter (+/- 2,859)
test search_index_seeds ... bench:      33,954 ns/iter (+/- 2,944)
test search_index_seeds ... bench:      34,207 ns/iter (+/- 1,287)

average: ~33958 ns/iter
```

With the suggested change:

```
test search_index_seeds ... bench:      34,387 ns/iter (+/- 3,337)
test search_index_seeds ... bench:      33,873 ns/iter (+/- 2,172)
test search_index_seeds ... bench:      33,947 ns/iter (+/- 3,346)

average: ~34069 ns/iter

with RUSTFLAGS="-C target-cpu=native"

test search_index_seeds ... bench:      34,067 ns/iter (+/- 2,697)
test search_index_seeds ... bench:      34,046 ns/iter (+/- 1,759)
test search_index_seeds ... bench:      33,992 ns/iter (+/- 3,879)

average: ~34035 ns/iter
```

The difference is miniscule (about 0.2-0.3%) between the average run times. However, the variance of the version you suggest is slightly higher on my machine (average variance of 1784 ns as PR'd vs. 2951 ns as suggested for a generic x86_64 arch, and 2363 ns vs 2778 ns for the `target-cpu=native` arch), suggesting that the branch may actually have more predictable performance, at least on the random test data I included in that benchmark.

That said, I'm running these benches on a different machine than I had when I opened the PR, the Rust nightly version has changed, and I don't have time right now to dig into the generated assembly to see if there's anything wonky. 

My sense is that given a lack of clear evidence for the branchless version performing better, it would be better to opt for the version which more clearly communicates the intent of counting items which match our needle byte -- which I'd argue is the version currently submitted for review. That said, since it seems like a complete wash between the two versions (the 0.2% difference is well within the error bars of this measurement), I'm happy to switch it if you'd prefer.